### PR TITLE
do not attempt to havoc void-typed objects; fixes issue #2663

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/goto-programs/generate_function_bodies.cpp [hunk 1]
```diff
@@ -271,7 +271,8 @@ class havoc_generate_function_bodiest : public generate_function_bodiest,
       if(
         parameter.type().id() == ID_pointer &&
         std::regex_match(
-          id2string(parameter.get_base_name()), parameters_to_havoc))
+          id2string(parameter.get_base_name()), parameters_to_havoc) &&
+        parameter.type().subtype().id() != ID_empty)
```
Comment (tautschnig) [inline src/goto-programs/generate_function_bodies.cpp]: I think this should instead fall back to `char`, doing whatever would happen if the parameter were of type `char*`. Leaving the pointed-to object unmodified may fail to meet the user's expectations.

**Review by tautschnig — CHANGES_REQUESTED** [review]
See below; also lacks tests and I am not sure the commit message would make GitHub automatically close the issue.

File: src/goto-programs/generate_function_bodies.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-programs/generate_function_bodies.cpp]: I'd say that the user's expectation will be that the entire object that the 'void *' points to gets havoc'ed. 'char' seems arbitrary.

Comment (tautschnig) [reply src/goto-programs/generate_function_bodies.cpp]: So should we just refuse to generate code? Most cases using `void *` as parameter type will accept more than a single object type (e.g., arrays of varying length).

**Comment by tautschnig** [comment]
After some further discussion with @polgreen: in the same way that void pointers are skipped, so should be function pointers.

**Comment by kroening** [comment]
Indeed, done.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d02cf3044fb9b10209ad1d7ca2e1e6e5d2622683).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85620044

**Comment by NlightNFotis** [comment]
Hi @kroening, is this PR still relevant? If yes, can we get it rebased on `develop`? If not, can we get it closed?

**Comment by TGWDB** [comment]
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.

**Pull request closed without merging**